src/playground-code-editor.ts

```diff
@@ -384,17 +411,252 @@ export class PlaygroundCodeEditor extends LitElement {
         this._applyHideAndFoldRegions();
         this._showDiagnostics();
       } else {
-        // Change event is only for user input.
         this.dispatchEvent(new Event('change'));
+        this._requestCompletionsIfNeeded(cm, changeObject);
       }
     });
     this._codemirror = cm;
   }
 
+  private _requestCompletionsIfNeeded(
+    editorInstance: Editor,
+    changeObject: EditorChange
+  ) {
+    if (!this._currentFiletypeSupportsCompletion()) return;
+
+    const previousToken = editorInstance.getTokenAt(changeObject.from);
+    const tokenUnderCursor = this.tokenUnderCursor.string.trim();
+    const tokenUnderCursorAsString = tokenUnderCursor.trim();
+    // To help reduce round trips to a language service or a completion provider, we
+    // are providing a flag if the completion is building on top of the earlier recommendations.
+    // If the flag is true, the completion system can just filter the already stored
+    // collection of completions again with the more precise input.
+    // On deletion events, we want to query the LS again, since we might be in a new context after
+    // removing characters from our code.
+    const isInputEvent = changeObject.origin === '+input';
+    const isRefinement =
+      (tokenUnderCursor.length > 1 || previousToken.string === '.') &&
+      isInputEvent;
+    const changeWasCodeCompletion = changeObject.origin === 'complete';
+
+    if (tokenUnderCursorAsString.length <= 0) return;
+
+    if (changeWasCodeCompletion) {
+      // If the case that the user triggered a code completion,
+      // we want to empty out the completions until
+      // a letter is input.
+      this.completions = [];
+      return;
+    }
+
+    const id = ++this._currentCompletionRequestId;
+    this.dispatchEvent(
+      new CustomEvent('request-completions', {
+        detail: {
+          isRefinement,
+          fileName: this.fileName,
+          fileContent: this.value,
+          tokenUnderCursor,
+          cursorIndex: this.cursorIndex,
+          provideCompletions: (completions: EditorCompletion[]) =>
+            this._onCompletionsProvided(id, completions),
+        },
+      })
+    );
+  }
+
+  private _onCompletionsProvided(id: number, completions: EditorCompletion[]) {
+    // To prevent race conditioning, check that the completions provided
+    // are from the latest completions request.
+    if (id !== this._currentCompletionRequestId) {
+      return;
+    }
+
+    this.completions = completions;
+  }
+
+  private _currentFiletypeSupportsCompletion() {
+    // Currently we are only supporting code completion for TS. Change
+    // this in a case that we start to support it for other languages too.
+    return this.type === 'ts';
+  }
+
   focus() {
     this._codemirrorEditable?.focus();
   }
 
+  private _hintFunction(
```

Comment:
How about renaming this to `_completionsAsHints`, and inlining the current `_completionsAsHints` function into this one. I think that would be easier to read, since they're very closely related.